components/brave_wallet/browser/buy_and_sell_response_parser.cc

```diff
@@ -0,0 +1,520 @@
+        sp_item.GetDict().FindString("serviceProvider");
+    if (!sp_service_provider) {
+      return false;
+    }
+    sp->service_provider = *sp_service_provider;
+
+    const std::string* sp_site_url = sp_item.GetDict().FindString("websiteUrl");
+    if (!sp_site_url) {
+      return false;
+    }
+    sp->web_site_url = *sp_site_url;
+
+    sp->logo_images = mojom::LogoImages::New();
+    if (const auto* logos = sp_item.GetDict().FindDict("logos");
+        !ParseMeldLogos(logos, sp->logo_images.get())) {
+      return false;
+    }
+
+    service_providers->emplace_back(std::move(sp));
+  }
+
+  return true;
+}
+
+bool ParseMeldErrorResponse(const base::Value& json_value,
+                            std::vector<std::string>* errors) {
+  // Parses results like this:
+  // {
+  //     "code": "BAD_REQUEST",
+  //     "message": "Bad request",
+  //     "errors": [
+  //         "[amount] Must be a decimal value greater than zero"
+  //     ],
+  //     "requestId": "eb6aaa76bd7103cf6c5b090610c31913",
+  //     "timestamp": "2022-01-19T20:32:30.784928Z"
+  // }
+  DCHECK(errors);
+  if (!json_value.is_dict()) {
+    return false;
+  }
+
+  const auto& response_error_dict = json_value.GetDict();
+
+  if (const auto* response_errors = response_error_dict.FindList("errors");
+      response_errors) {
+    for (const auto& err : *response_errors) {
+      errors->emplace_back(err.GetString());
+    }
+  }
+
+  if (const auto* response_error_message =
+          response_error_dict.FindString("message");
+      response_error_message && errors->empty()) {
+    errors->emplace_back(*response_error_message);
+  }
+
+  return !errors->empty();
+}
+
+bool ParseCryptoQuotes(const base::Value& json_value,
+                       std::vector<mojom::CryptoQuotePtr>* quotes,
+                       std::string* error) {
+  // Parses results like this:
+  // {
+  //   "quotes": [
+  //     {
+  //       "transactionType": "CRYPTO_PURCHASE",
+  //       "sourceAmount": 50,
+  //       "sourceAmountWithoutFees": 43.97,
+  //       "fiatAmountWithoutFees": 43.97,
+  //       "destinationAmountWithoutFees": null,
+  //       "sourceCurrencyCode": "USD",
+  //       "countryCode": "US",
+  //       "totalFee": 6.03,
+  //       "networkFee": 3.53,
+  //       "transactionFee": 2,
+  //       "destinationAmount": 0.00066413,
+  //       "destinationCurrencyCode": "BTC",
+  //       "exchangeRate": 75286,
+  //       "paymentMethodType": "APPLE_PAY",
+  //       "customerScore": 20,
+  //       "serviceProvider": "TRANSAK"
+  //     }
+  //   ],
+  //   "message": null,
+  //   "error": null
+  // }
+  DCHECK(quotes);
+  DCHECK(error);
+
+  if (!json_value.is_dict()) {
+    LOG(ERROR) << "Invalid response, could not parse JSON, JSON is not a dict";
+    return false;
+  }
+
+  const auto& response_dict = json_value.GetDict();
+  if (const auto* response_error = response_dict.FindString("error")) {
+    *error = *response_error;
+  }
+
```

Comment:
Should `ParseMeldErrorResponse` already capture error?

Reply:
It is not the same errors, for example `CryptoQuoteResponse` looks like:
```
{
  "quotes": [
    {
      "transactionType": "CRYPTO_PURCHASE",
      "sourceAmount": 50,
      "sourceAmountWithoutFees": 43.97,
      "fiatAmountWithoutFees": 43.97,
      "destinationAmountWithoutFees": null,
      "sourceCurrencyCode": "USD",
      "countryCode": "US",
      "totalFee": 6.03,
      "networkFee": 3.53,
      "transactionFee": 2,
      "destinationAmount": 0.00066413,
      "destinationCurrencyCode": "BTC",
      "exchangeRate": 75286,
      "paymentMethodType": "APPLE_PAY",
      "customerScore": 20,
      "serviceProvider": "TRANSAK"
    }
  ],
  "message": null,
  "error": null
}
```
with that code we parse property: `"error": null`